File: frontend/src/api/playlists.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query"

import { fetchJson } from "@/api/client"
import type { PlaylistListItem, RebuildStatus, RefreshCadence, WritebackStatus } from "@/api/types"

const WEEKDAY_NAMES = [
  "Monday",
  "Tuesday",
  "Wednesday",
  "Thursday",
  "Friday",
  "Saturday",
  "Sunday",
] as const

// ── Extra domain types ─────────────────────────────────────────────────────

export type RowMode = "ordered" | "disordered"
export type CompletionPolicy = "remove" | "restart" | "disordered"
export type SlotAllocation = "wild" | "balanced" | "round_robin"

export const SLOT_ALLOCATION_LABELS: Record<SlotAllocation, string> = {
  wild: "Wild",
  balanced: "Balanced",
  round_robin: "Round-robin",
}

export interface SnapshotEpisode {
  episode_id: string
  title: string
  series_id: string
  series_title: string | null
  slot_index: number
  row_mode: string
}

export interface PlaylistSeriesRowResponse {
  series_id: string
  mode: RowMode
  completion_policy: CompletionPolicy
  completion_event: string
  series_title: string | null
  thumb_url: string | null
}

export interface RebuildRunSummary {
  id: string
  status: string
  started_at: string | null
  finished_at: string | null
  error_message: string | null
  slots_filled: number | null
  slots_requested: number | null
  writeback_status?: WritebackStatus
  writeback_error?: string | null
  writeback_warnings?: Array<{ episode_id?: string | null; reason?: string }> | null
  writeback_at?: string | null
}

export interface PlaylistDetailResponse {
  id: string
  name: string
  episode_count: number
  slot_allocation: SlotAllocation
  default_completion_policy: CompletionPolicy
  refresh_cadence: RefreshCadence
  refresh_day_of_week: number | null
  rows: PlaylistSeriesRowResponse[]
  current_snapshot: SnapshotEpisode[]
  last_rebuild: RebuildRunSummary | null
  recent_runs: RebuildRunSummary[]
  provider_playlist_id?: string | null
  provider_kind?: string | null
  provider_playlist_open_url?: string | null
}

export interface PlaylistSeriesRowPayload {
  series_id: string
  mode: RowMode
  completion_policy: CompletionPolicy
}

export interface PlaylistCreatePayload {
  name: string
  episode_count: number
  slot_allocation: SlotAllocation
  default_completion_policy: CompletionPolicy
  refresh_cadence: RefreshCadence
  refresh_day_of_week: number | null
  rows: PlaylistSeriesRowPayload[]
}

export type PlaylistUpdatePayload = Partial<PlaylistCreatePayload>

// ── Fetch functions ────────────────────────────────────────────────────────

export async function fetchPlaylists(seriesId?: string): Promise<PlaylistListItem[]> {
  const query = seriesId
    ? `?series_id=${encodeURIComponent(seriesId)}`
    : ""
  return fetchJson<PlaylistListItem[]>(`/playlists${query}`)
}

export async function fetchPlaylist(id: string): Promise<PlaylistDetailResponse> {
  return fetchJson<PlaylistDetailResponse>(`/playlists/${id}`)
}

export async function createPlaylist(payload: PlaylistCreatePayload): Promise<PlaylistDetailResponse> {
  return fetchJson<PlaylistDetailResponse>("/playlists", {
    method: "POST",
    body: JSON.stringify(payload),
  })
}

export async function updatePlaylist(id: string, payload: PlaylistUpdatePayload): Promise<PlaylistDetailResponse> {
  return fetchJson<PlaylistDetailResponse>(`/playlists/${id}`, {
    method: "PUT",
    body: JSON.stringify(payload),
  })
}

export async function deletePlaylist(id: string): Promise<void> {
  await fetchJson<void>(`/playlists/${id}`, { method: "DELETE" })
}

export async function triggerRebuild(id: string): Promise<RebuildRunSummary> {
  return fetchJson<RebuildRunSummary>(`/playlists/${id}/rebuild`, { method: "POST" })
}

export interface AppendPlaylistRowPayload {
  series_id: string
  mode?: RowMode
  completion_policy?: CompletionPolicy
}

export async function appendPlaylistRow(
  playlistId: string,
  payload: AppendPlaylistRowPayload,
): Promise<PlaylistDetailResponse> {
  return fetchJson<PlaylistDetailResponse>(`/playlists/${playlistId}/rows`, {
    method: "POST",
    body: JSON.stringify(payload),
  })
}

export interface PatchPlaylistRowPayload {
  mode?: RowMode
  completion_policy?: CompletionPolicy
}

export async function removePlaylistRow(
  playlistId: string,
  seriesId: string,
): Promise<void> {
  await fetchJson<void>(
    `/playlists/${playlistId}/rows/${encodeURIComponent(seriesId)}`,
    { method: "DELETE" },
  )
}

export async function patchPlaylistRow(
  playlistId: string,
  seriesId: string,
  payload: PatchPlaylistRowPayload,
): Promise<PlaylistDetailResponse> {
  return fetchJson<PlaylistDetailResponse>(
    `/playlists/${playlistId}/rows/${encodeURIComponent(seriesId)}`,
    { method: "PATCH", body: JSON.stringify(payload) },
  )
}

export async function createPlaylistWithSeries(
  name: string,
  seriesId: string,
): Promise<PlaylistDetailResponse> {
  return createPlaylist({
    name: name.trim(),
    episode_count: 20,
    slot_allocation: "wild",
    default_completion_policy: "remove",
    refresh_cadence: "daily",
    refresh_day_of_week: null,
    rows: [
      {
        series_id: seriesId,
        mode: "ordered",
        completion_policy: "remove",
      },
    ],
  })
}

// ── Query hooks ────────────────────────────────────────────────────────────

export function usePlaylists() {
  return useQuery({
    queryKey: ["playlists", "list"],
    queryFn: () => fetchPlaylists(),
    staleTime: 30_000,
  })
}

export function usePlaylistsContainingSeries(seriesId: string | undefined) {
  return useQuery({
    queryKey: ["playlists", "containing", seriesId],
    queryFn: () => fetchPlaylists(seriesId!),
    enabled: Boolean(seriesId),
    staleTime: 30_000,
  })
}

const POLLING_STATUSES: RebuildStatus[] = ["running", "queued"]

export function usePlaylist(id: string) {
  return useQuery({
    queryKey: ["playlists", id],
    queryFn: () => fetchPlaylist(id),
    staleTime: 5_000,
    refetchInterval: (query) => {
      const status = query.state.data?.last_rebuild?.status as RebuildStatus | undefined
      return status && POLLING_STATUSES.includes(status) ? 5_000 : false
    },
  })
}

// ── Mutation hooks ─────────────────────────────────────────────────────────

export function useCreatePlaylist() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: createPlaylist,
    onSuccess: () => {
      void queryClient.invalidateQueries({ queryKey: ["playlists"] })
    },
  })
}

export function useUpdatePlaylist() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({ id, payload }: { id: string; payload: PlaylistUpdatePayload }) =>
      updatePlaylist(id, payload),
    onSuccess: (_data, { id }) => {
      void queryClient.invalidateQueries({ queryKey: ["playlists"] })
      void queryClient.invalidateQueries({ queryKey: ["playlists", id] })
    },
  })
}

export function useDeletePlaylist() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: deletePlaylist,
    onSuccess: () => {
      void queryClient.invalidateQueries({ queryKey: ["playlists"] })
    },
  })
}

export function useRebuildPlaylist() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: triggerRebuild,
    onSuccess: (_data, id) => {
      void queryClient.invalidateQueries({ queryKey: ["playlists", id] })
    },
  })
}

export function useAppendPlaylistRow() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({
      playlistId,
      payload,
    }: {
      playlistId: string
      payload: AppendPlaylistRowPayload
    }) => appendPlaylistRow(playlistId, payload),
    onSuccess: (_data, { playlistId }) => {
      void queryClient.invalidateQueries({ queryKey: ["playlists", playlistId] })
      void queryClient.invalidateQueries({ queryKey: ["playlists"] })
    },
  })
}

export function useRemovePlaylistRow() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({
      playlistId,
      seriesId,
    }: {
      playlistId: string
      seriesId: string
    }) => removePlaylistRow(playlistId, seriesId),
    onSuccess: (_data, { playlistId }) => {
      void queryClient.invalidateQueries({ queryKey: ["playlists", playlistId] })
      void queryClient.invalidateQueries({ queryKey: ["playlists"] })
    },
  })
}

export function usePatchPlaylistRow() {
  const queryClient = useQueryClient()
  return useMutation({
    mutationFn: ({
      playlistId,
      seriesId,
      payload,
    }: {
      playlistId: string
      seriesId: string
      payload: PatchPlaylistRowPayload
    }) => patchPlaylistRow(playlistId, seriesId, payload),
    onSuccess: (_data, { playlistId }) => {
      void queryClient.invalidateQueries({ queryKey: ["playlists", playlistId] })
      void queryClient.invalidateQueries({ queryKey: ["playlists"] })
    },
  })
}

// ── Formatters ─────────────────────────────────────────────────────────────

export function formatCadence(item: {
  refresh_cadence: RefreshCadence
  refresh_day_of_week: number | null
}): string {
  if (item.refresh_cadence === "daily") {
    return "Daily"
  }
  const dayName =
    item.refresh_day_of_week != null
      ? WEEKDAY_NAMES[item.refresh_day_of_week]
      : null
  return dayName ? `Weekly \u00b7 ${dayName}` : "Weekly"
}

File: backend/src/wheeloffish/api/schemas/playlists.py
"""Pydantic request/response schemas for playlist CRUD + rebuild API (Phase 05 Plan 04)."""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, model_validator

from wheeloffish.domain.playlist import (
    CompletionEvent,
    CompletionPolicy,
    RowMode,
    SlotAllocation,
)


class PlaylistSeriesRowRequest(BaseModel):
    series_id: str
    mode: RowMode = RowMode.ORDERED
    completion_policy: CompletionPolicy = CompletionPolicy.REMOVE
    completion_event: CompletionEvent = CompletionEvent.SERIES_COMPLETE


class PlaylistCreateRequest(BaseModel):
    name: str = Field(..., min_length=1)
    episode_count: int = Field(default=20, ge=1)
    slot_allocation: SlotAllocation = SlotAllocation.WILD
    default_completion_policy: CompletionPolicy = CompletionPolicy.REMOVE
    refresh_cadence: str = Field(default="daily", pattern=r"^(daily|weekly)$")
    refresh_day_of_week: int | None = Field(default=None, ge=0, le=6)
    rows: list[PlaylistSeriesRowRequest] = Field(default_factory=list)

    @model_validator(mode="after")
    def weekly_requires_day(self) -> PlaylistCreateRequest:
        if self.refresh_cadence == "weekly" and self.refresh_day_of_week is None:
            raise ValueError("refresh_day_of_week required when refresh_cadence is weekly")
        return self


class PlaylistUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1)
    episode_count: int | None = Field(default=None, ge=1)
    slot_allocation: SlotAllocation | None = None
    default_completion_policy: CompletionPolicy | None = None
    refresh_cadence: str | None = Field(default=None, pattern=r"^(daily|weekly)$")
    refresh_day_of_week: int | None = Field(default=None, ge=0, le=6)
    rows: list[PlaylistSeriesRowRequest] | None = None

    @model_validator(mode="after")
    def weekly_requires_day(self) -> PlaylistUpdateRequest:
        if self.refresh_cadence == "weekly" and self.refresh_day_of_week is None:
            raise ValueError("refresh_day_of_week required when refresh_cadence is weekly")
        return self


class AppendRowRequest(BaseModel):
    series_id: str = Field(..., min_length=1)
    mode: RowMode = RowMode.ORDERED
    completion_policy: CompletionPolicy = CompletionPolicy.REMOVE
    completion_event: CompletionEvent = CompletionEvent.SERIES_COMPLETE


class PatchRowRequest(BaseModel):
    mode: RowMode | None = None
    completion_policy: CompletionPolicy | None = None
    completion_event: CompletionEvent | None = None

    @model_validator(mode="after")
    def at_least_one_field(self) -> PatchRowRequest:
        if (
            self.mode is None
            and self.completion_policy is None
            and self.completion_event is None
        ):
            raise ValueError(
                "At least one of mode, completion_policy, or completion_event required"
            )
        return self


class DiagnosticAction(BaseModel):
    type: Literal["remove_row", "open_provider", "open_series"]
    label: str
    series_id: str | None = None
    episode_id: str | None = None
    url: str | None = None


class DiagnosticIssueRow(BaseModel):
    label: str
    reason_code: str
    reason_text: str
    remediation_hint: str
    series_id: str | None = None
    episode_id: str | None = None
    actions: list[DiagnosticAction] = Field(default_factory=list)


class RebuildDiagnostics(BaseModel):
    rebuild_error: DiagnosticIssueRow | None = None
    show_issues: list[DiagnosticIssueRow] = Field(default_factory=list)
    episode_issues: list[DiagnosticIssueRow] = Field(default_factory=list)


class RebuildRunSummary(BaseModel):
    id: str
    status: str
    started_at: datetime | None
    finished_at: datetime | None
    error_message: str | None = None
    slots_filled: int | None = None
    slots_requested: int | None = None
    writeback_status: str | None = None
    writeback_error: str | None = None
    writeback_warnings: list[dict] | None = None
    writeback_at: datetime | None = None
    diagnostics: RebuildDiagnostics | None = None


class PruneEventResponse(BaseModel):
    id: str
    series_id: str
    event_type: str
    reason: str
    event_metadata: dict | None = None
    timestamp: datetime


class SnapshotEpisode(BaseModel):
    episode_id: str
    title: str
    series_id: str
    series_title: str | None = None
    slot_index: int
    row_mode: str


class PlaylistSeriesRowResponse(BaseModel):
    series_id: str
    mode: str
    completion_policy: str
    completion_event: str
    series_title: str | None = None
    thumb_url: str | None = None


class PlaylistDetailResponse(BaseModel):
    id: str
    name: str
    episode_count: int
    slot_allocation: str
    default_completion_policy: str
    refresh_cadence: str
    refresh_day_of_week: int | None
    rows: list[PlaylistSeriesRowResponse]
    current_snapshot: list[SnapshotEpisode]
    last_rebuild: RebuildRunSummary | None
    recent_runs: list[RebuildRunSummary]
    recent_prune_events: list[PruneEventResponse] = Field(default_factory=list)
    provider_playlist_id: str | None = None
    provider_kind: str | None = None
    provider_playlist_open_url: str | None = None


class PlaylistListItem(BaseModel):
    id: str
    name: str
    refresh_cadence: str
    refresh_day_of_week: int | None
    last_rebuild_status: str | None
    last_rebuild_at: datetime | None
    last_writeback_status: str | None = None
    provider_playlist_id: str | None = None
    provider_kind: str | None = None
    provider_playlist_open_url: str | None = None
